src/config.rs

```diff
@@ -120,106 +220,388 @@ impl<'a> Deserialize<'a> for UserNodeConfig {
             for input in &default_inputs {
-                graph.add(input, name)
+                links.push(UserLink {
+                    from: UserNodePort {
+                        node: Some(input.other_node.clone()),
+                        port: Some(input.other_port.clone()),
+                    },
+                    to: UserNodePort {
+                        node: Some(name.into()),
+                        port: Some(input.this_port.clone()),
+                    },
+                });
             }
         }
     }
-    if unc.outputs.is_empty() {
+    if n_outputs == 0 {
         if let Some(default_outputs) = nc.default_outputs() {
             for output in &default_outputs {
-                graph.add(name, output)
+                links.push(UserLink {
+                    from: UserNodePort {
+                        node: Some(name.into()),
+                        port: Some(output.this_port.clone()),
+                    },
+                    to: UserNodePort {
+                        node: Some(output.other_node.clone()),
+                        port: Some(output.other_port.clone()),
+                    },
+                });
             }
         }
     }
 }
 
-impl Config {
-    pub fn new(config_bytes: Vec<u8>) -> Result<Config, String> {
-        match de::from_slice::<UserConfig>(&config_bytes) {
-            Ok(user_config) => {
-                let mut node_list = Vec::new();
-                let mut node_names = Vec::new();
-                let mut graph: DependencyGraph = Default::default();
+fn push_ports(ports: &mut Vec<Vec<String>>, pc: PortConfig, given: &Vec<String>) -> bool {
+    let mut list = pc.defaults.unwrap_or_default();
 
-                for unc in &user_config.nodes {
-                    let name: &str = &unc.name;
+    if pc.user_defined_ports {
+        for port in given {
+            if !list.iter().any(|p| p == port) {
+                list.push(port.into());
+            }
+        }
+    }
 
-                    if RESERVED_NODE_NAMES.contains(name) {
-                        return Err(format!("cannot use reserved node name '{name}'"));
-                    }
+    ports.push(list);
 
-                    node_names.push(name.to_string());
-                    for input in &unc.inputs {
-                        graph.add(input, name);
-                    }
-                    for output in &unc.outputs {
-                        graph.add(name, output);
-                    }
+    pc.user_defined_ports
+}
+
+fn has(xs: &[String], s: &str) -> bool {
+    xs.iter().any(|x| x == s)
+}
+
+fn resolve_port_names(
+    link: &mut UserLink,
+    outs: &mut Vec<String>,
+    user_outs: bool,
+    ins: &mut Vec<String>,
+    user_ins: bool,
+    n_linked_inputs: usize,
+) -> Result<(), String> {
+    let mut from_port = None;
+    let mut to_port = None;
+    match &link.from.port {
+        Some(port) => {
+            if !has(outs, port) {
```

Comment:
I think there's something in stdlib for this

```suggestion
            if !outs.contains(port) {
```

Reply:
I was able to replace all but one (which used Vec<String> and &str), so I just used .iter().any() as recommended by the manual: https://doc.rust-lang.org/std/primitive.slice.html#method.contains